openssl/openssl.go

```diff
@@ -258,3 +258,28 @@ func bnToBig(bn C.GO_BIGNUM_PTR) *big.Int {
 	n := C.go_openssl_BN_bn2bin(bn, base(raw))
 	return new(big.Int).SetBytes(raw[:n])
 }
+
+// noescape hides a pointer from escape analysis. noescape is
+// the identity function but escape analysis doesn't think the
+// output depends on the input. noescape is inlined and currently
+// compiles down to zero instructions.
+// USE CAREFULLY!
+//go:nosplit
+func noescape(p unsafe.Pointer) unsafe.Pointer {
```

Comment:
I read up a bit about this. Escape logic impacts cgo in the sense that it's important that pointers which are passed to C **and** are stored there need to be allocated on the heap. If they are allocated on the stack then that creates a memory hole. The default of cgo is to ensure all pointers are marked as escaping so they get pushed to the heap (based on the Russ change). This is reasonable because it's the safest default. 

The `noescape` function is then used to trick the cgo tooling. Basically and "I know better, this isn't stored" escape hatch for developers

This is my reading of it  at least

Reply:
I have the same sense based on https://github.com/golang/go/issues/10303.

https://github.com/golang/go/issues/24450 also points out that even though cgo has some strict-seeming pointer passing prerequisites, it's still not always easy to identify when it's safe to noescape. (If the C function calls back into Go, Go might rearrange the stack, so even having a pointer to something in the Go stack in a C parameter or local variable can turn out badly.)

Reply:
To the best of my knowledge, OpenSSL is not storing the few pointers we are passing as `noescape`, so it is safe to trick cgo tooling.